pkg/instancetype/controller/vm/controller_test.go

```diff
@@ -0,0 +1,1052 @@
+//nolint:dupl
+package vm_test
+
+import (
+	"context"
+	"encoding/json"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/golang/mock/gomock"
+	appsv1 "k8s.io/api/apps/v1"
+	k8sv1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	"kubevirt.io/api/instancetype/v1alpha1"
+	"kubevirt.io/api/instancetype/v1alpha2"
+	"kubevirt.io/api/instancetype/v1beta1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/kubevirt/fake"
+
+	"kubevirt.io/kubevirt/pkg/instancetype"
+	instancetypecontroller "kubevirt.io/kubevirt/pkg/instancetype/controller/vm"
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/pkg/pointer"
+	"kubevirt.io/kubevirt/pkg/testutils"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	"kubevirt.io/kubevirt/pkg/virt-config/featuregate"
+	"kubevirt.io/kubevirt/pkg/virt-controller/watch/common"
+)
+
+var _ = Describe("Instance type and Preference VirtualMachine Controller", func() {
+	const (
+		resourceUID        types.UID = "9160e5de-2540-476a-86d9-af0081aee68a"
+		resourceGeneration int64     = 1
+	)
+
+	type instancetypeVMController interface {
+		Sync(*virtv1.VirtualMachine, *virtv1.VirtualMachineInstance) (*virtv1.VirtualMachine, error)
+		ApplyToVM(*virtv1.VirtualMachine) error
+		ApplyToVMI(*virtv1.VirtualMachine, *virtv1.VirtualMachineInstance) error
+		ApplyDevicePreferences(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error
+	}
+
+	var (
+		vm  *virtv1.VirtualMachine
+		vmi *virtv1.VirtualMachineInstance
+
+		instancetypeController instancetypeVMController
+
+		instancetypeObj *v1beta1.VirtualMachineInstancetype
+		preference      *v1beta1.VirtualMachinePreference
+
+		config   *virtconfig.ClusterConfig
+		recorder *record.FakeRecorder
+
+		virtClient *kubecli.MockKubevirtClient
+
+		kvStore                          cache.Store
+		instancetypeInformerStore        cache.Store
+		clusterInstancetypeInformerStore cache.Store
+		preferenceInformerStore          cache.Store
+		clusterPreferenceInformerStore   cache.Store
+		controllerrevisionInformerStore  cache.Store
+	)
+
+	BeforeEach(func() {
+		vmi = libvmi.New(libvmi.WithNamespace(k8sv1.NamespaceDefault))
+		vm = libvmi.NewVirtualMachine(vmi)
+
+		// We need to clear the domainSpec here to ensure the instancetype doesn't conflict
+		vm.Spec.Template.Spec.Domain = virtv1.DomainSpec{}
+
+		ctrl := gomock.NewController(GinkgoT())
+		virtClient = kubecli.NewMockKubevirtClient(ctrl)
```

Comment:
Can you use the fake client directly instead of wrapping it in a mock client?

Reply:
As discussed there's no easy way of merging multiple fakeclients into a single thing we can pass on to the consumer here so the mockclient is used to facilitate this.